File: integration/DovahLinkValidationClient/RenameOutcomePayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Incoming <c>rename_outcome</c> payload (<c>protocol/schema/README.md</c>'s <c>rename_outcome</c>).
/// Decode-only: this client never sends <c>rename_outcome</c>.
/// </summary>
/// <param name="Outcome">The wire vocabulary of <c>rename_outcome.outcome</c>: <c>"renamed"</c>,
/// <c>"invalid_display_name"</c>, or <c>"not_trusted"</c>.</param>
/// <param name="DisplayName">Echoes the resulting name, present only for <c>"renamed"</c>;
/// <see langword="null"/> otherwise.</param>
public sealed record RenameOutcomePayload(string Outcome, string? DisplayName)
{
    /// <summary>
    /// Decodes and validates one <c>rename_outcome</c> payload.
    /// </summary>
    /// <param name="payload">The envelope's decoded payload object.</param>
    /// <returns>The decoded rename-outcome payload.</returns>
    /// <exception cref="FormatException">Thrown when a required key is missing or a value has the
    /// wrong JSON type.</exception>
    public static RenameOutcomePayload Decode(JsonObject payload)
    {
        try
        {
            string outcome = payload["outcome"]?.GetValue<string>() ?? throw new FormatException("Missing outcome.");
            if (!payload.ContainsKey("displayName"))
            {
                throw new FormatException("Missing displayName.");
            }
            string? displayName = payload["displayName"]?.GetValue<string>();
            return new RenameOutcomePayload(outcome, displayName);
        }
        catch (InvalidOperationException ex)
        {
            throw new FormatException($"Malformed rename_outcome payload: {ex.Message}", ex);
        }
    }
}

File: sdk/dart/dovahlink_client/lib/src/protocol/rename_outcome_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';
import 'package:dovahlink_client_sdk/src/protocol/protocol_format_exception.dart';
import 'package:dovahlink_client_sdk/src/shared/enums.dart';

part 'rename_outcome_payload.g.dart';

/// Incoming `rename_outcome` payload (`protocol/schema/README.md`'s `rename_outcome`). Decode-only:
/// the client never sends `rename_outcome`.
@JsonSerializable(checked: true, createToJson: false)
class RenameOutcomePayload {
  /// The wire vocabulary of `rename_outcome.outcome`.
  @JsonKey(required: true)
  final RenameOutcome outcome;

  /// Echoes the resulting name, present only for [RenameOutcome.renamed]; `null` otherwise.
  @JsonKey(required: true)
  final String? displayName;

  /// Creates a rename-outcome payload.
  const RenameOutcomePayload({
    required this.outcome,
    required this.displayName,
  });

  /// Decodes and validates one `rename_outcome` payload.
  factory RenameOutcomePayload.fromJson(JsonMap json) {
    try {
      final RenameOutcomePayload payload = _$RenameOutcomePayloadFromJson(json);
      if (payload.outcome != RenameOutcome.renamed &&
          payload.displayName != null) {
        throw const ProtocolFormatException(
          'displayName must be null unless outcome is renamed.',
        );
      }
      return payload;
    } on Object catch (error) {
      throw ProtocolFormatException('Invalid rename_outcome payload: $error');
    }
  }
}
